### DISCUSSION TITLE
Tutorial: “How to Install Coldtype from Scratch”

### DISCUSSION THREAD

**Discussion by stenson** [coldtype/coldtype#60] [Videos]
https://www.youtube.com/watch?v=hCGF6FZCSqc

**Comment by mgpjm** [comment]
Hi, while installing Coldtype viewer I run into this error, and I can't figure it out how I can solve this (MacOS Big Sur 11.2.3):
`  Running setup.py install for noise ... error`
` ERROR: Command errored out with exit status 1:
     command: /Users/mgpjm/ColdType/Tutorials/installing-coldtype/venv/bin/python3.9 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/t2/nm2b4yr15sbfqdx_jg60fxw80000gn/T/pip-install-yei0bl71/noise_19984a7f1db94feebff2daa48cfa2a0c/setup.py'"'"'; __file__='"'"'/private/var/folders/t2/nm2b4yr15sbfqdx_jg60fxw80000gn/T/pip-install-yei0bl71/noise_19984a7f1db94feebff2daa48cfa2a0c/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/folders/t2/nm2b4yr15sbfqdx_jg60fxw80000gn/T/pip-record-nrjkv9p8/install-record.txt --single-version-externally-managed --compile --install-headers /Users/mgpjm/ColdType/Tutorials/installing-coldtype/venv/include/site/python3.9/noise
         cwd: /private/var/folders/t2/nm2b4yr15sbfqdx_jg60fxw80000gn/T/pip-install-yei0bl71/noise_19984a7f1db94feebff2daa48cfa2a0c/
    Complete output (17 lines):
    running install
    running build
    running build_py
    creating build
    creating build/lib.macosx-10.9-x86_64-3.9
    creating build/lib.macosx-10.9-x86_64-3.9/noise
    copying shader.py -> build/lib.macosx-10.9-x86_64-3.9/noise
    copying perlin.py -> build/lib.macosx-10.9-x86_64-3.9/noise
    copying shader_noise.py -> build/lib.macosx-10.9-x86_64-3.9/noise
    copying __init__.py -> build/lib.macosx-10.9-x86_64-3.9/noise
    copying test.py -> build/lib.macosx-10.9-x86_64-3.9/noise
    running build_ext
    building 'noise._simplex' extension
    creating build/temp.macosx-10.9-x86_64-3.9
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -arch x86_64 -g -I/Users/mgpjm/ColdType/Tutorials/installing-coldtype/venv/include -I/Library/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c _simplex.c -o build/temp.macosx-10.9-x86_64-3.9/_simplex.o -funroll-loops
    xcrun: error: invalid active developer path (/Library/Developer/CommandLineTools), missing xcrun at: /Library/Developer/CommandLineTools/usr/bin/xcrun
    error: command '/usr/bin/gcc' failed with exit code 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /Users/mgpjm/ColdType/Tutorials/installing-coldtype/venv/bin/python3.9 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/t2/nm2b4yr15sbfqdx_jg60fxw80000gn/T/pip-install-yei0bl71/noise_19984a7f1db94feebff2daa48cfa2a0c/setup.py'"'"'; __file__='"'"'/private/var/folders/t2/nm2b4yr15sbfqdx_jg60fxw80000gn/T/pip-install-yei0bl71/noise_19984a7f1db94feebff2daa48cfa2a0c/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/folders/t2/nm2b4yr15sbfqdx_jg60fxw80000gn/T/pip-record-nrjkv9p8/install-record.txt --single-version-externally-managed --compile --install-headers /Users/mgpjm/ColdType/Tutorials/installing-coldtype/venv/include/site/python3.9/noise Check the logs for full command output.`

**Comment by stenson** [comment]
@mgpjm thanks for trying it out, and apologies for the issue! It looks like your computer isn't able to compile the `noise` library (probably because of an issue with your Xcode / Xcode command line tools installation), but I've just pushed a new version of coldtype (0.7.0) that no longer requires the `noise` package. Could you try running `pip install coldtype[viewer] --upgrade` and see if that works?

**Reply by mgpjm** [reply]
Yeee it works now! 
Thank you @stenson

**Comment by ghost** [comment]
Hi, I guess I'm encountering sorta similar issue when trying to install on macOS 11.6.1 (tried two separate partitions):

    Running setup.py install for unicodedata2 ... error
    ERROR: Command errored out with exit status 1:
     command: /Users/cookie/Tutorials/installing-coldtype/venv/bin/python3.10 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/xy/dz4x1k7d56l4gpbd362zb3m40000gn/T/pip-install-67t6md73/unicodedata2_153c77f57eb54bdcb8c09ab74d136323/setup.py'"'"'; __file__='"'"'/private/var/folders/xy/dz4x1k7d56l4gpbd362zb3m40000gn/T/pip-install-67t6md73/unicodedata2_153c77f57eb54bdcb8c09ab74d136323/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/folders/xy/dz4x1k7d56l4gpbd362zb3m40000gn/T/pip-record-curo1qlj/install-record.txt --single-version-externally-managed --compile --install-headers /Users/cookie/Tutorials/installing-coldtype/venv/include/site/python3.10/unicodedata2
         cwd: /private/var/folders/xy/dz4x1k7d56l4gpbd362zb3m40000gn/T/pip-install-67t6md73/unicodedata2_153c77f57eb54bdcb8c09ab74d136323/
    Complete output (16 lines):
    /Users/cookie/Tutorials/installing-coldtype/venv/lib/python3.10/site-packages/setuptools/dist.py:484: UserWarning: Normalizing '13.0.0-2' to '13.0.0.post2'
      warnings.warn(tmpl.format(**locals()))
    running install
    running build
    running build_ext
    building 'unicodedata2' extension
    creating build
    creating build/temp.macosx-10.9-universal2-3.10
    creating build/temp.macosx-10.9-universal2-3.10/unicodedata2
    creating build/temp.macosx-10.9-universal2-3.10/unicodedata2/py3
    clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -arch arm64 -arch x86_64 -g -I./unicodedata2/py3 -I./unicodedata2/ -I/Users/cookie/Tutorials/installing-coldtype/venv/include -I/Library/Frameworks/Python.framework/Versions/3.10/include/python3.10 -c ./unicodedata2/py3/unicodedata.c -o build/temp.macosx-10.9-universal2-3.10/./unicodedata2/py3/unicodedata.o
    ./unicodedata2/py3/unicodedata.c:19:10: fatal error: 'ucnhash.h' file not found
    #include "ucnhash.h"
             ^~~~~~~~~~~
    1 error generated.
    error: command '/usr/bin/clang' failed with exit code 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /Users/cookie/Tutorials/installing-coldtype/venv/bin/python3.10 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/xy/dz4x1k7d56l4gpbd362zb3m40000gn/T/pip-install-67t6md73/unicodedata2_153c77f57eb54bdcb8c09ab74d136323/setup.py'"'"'; __file__='"'"'/private/var/folders/xy/dz4x1k7d56l4gpbd362zb3m40000gn/T/pip-install-67t6md73/unicodedata2_153c77f57eb54bdcb8c09ab74d136323/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/folders/xy/dz4x1k7d56l4gpbd362zb3m40000gn/T/pip-record-curo1qlj/install-record.txt --single-version-externally-managed --compile --install-headers /Users/cookie/Tutorials/installing-coldtype/venv/include/site/python3.10/unicodedata2 Check the logs for full command output.

I've tried both pip install "coldtype[viewer]" and pip install "coldtype[viewer] --upgrade", but still getting stuck with the same error.
Any help?

**Reply by stenson** [reply]
Hi @fishermansblues, thanks for trying out Coldtype! Python 3.10 wasn't out when I made that tutorial, and unfortunately 3.10 doesn't yet work with all the required coldtype dependencies (chiefly `skia-python`, which is a core dependency).

For best results as of this writing, I'd recommend Python 3.9.8 intel-only download available here: https://www.python.org/downloads/macos/ (and more specifically here: https://www.python.org/downloads/release/python-398/), and then you can address that in your terminal as python

I'll be looking into solutions for Python 3.10 but for the moment it's a bit of a dead-end if you're looking to follow along with that video.

**Reply by ghost** [reply]
Wow, thanks for the prompt response, and guidance! Will give 3.9.8 a try soon!

**Reply by ghost** [reply]
That worked! Will continue with the tutorial vid. Thanks again!

**Reply by stenson** [reply]
Great to hear!

**Reply by ghost** [reply]
Sorry to be a pain in the butt on Saturday.

I worked through the tutorials, but the only part I get stuck with is:
    wdth=f.e("qeio", 1),
    wght=f.e("seio", 2),
    slnt=1)
The test.py doesn't seem to react to the above code.

**Reply by stenson** [reply]
no worries! just seeing this now — what font are you using? it could be that you're using a font that doesn't support those axes. You could try `Font.MutatorSans()` also, which will definitely support `wdth` and `wght`

**Reply by ghost** [reply]
I was using Obviously Demo fonts, but it looks like MutatorSans is working! Thanks again!

**Reply by stenson** [reply]
Great! Ah yeah that makes sense, demo versions of fonts usually don't include the variable versions, both since variable fonts are still kind of an emerging technology and because they're harder to pare down to a demo version. In general, if the font you're trying to use has an ".otf" file type, then it won't be variable. A ".ttf" doesn't guarantee that it will be variable, but that's the most common file type for variable fonts. If you are looking to mess around with variable fonts, you can find a whole bunch of for-sale (and some free) ones on https://v-fonts.com/

**Reply by ghost** [reply]
Ahhhh now I know! Thanks for being super helpful!

**Comment by mgpjm** [comment]
Hi Rob, 
I'm trying to install Coldtype on a new machine (macOS Monterey M1) getting a new kind of error while I try to run Coldtype Demo

```
"File "/Users/mgpjm/animation/venv/bin/coldtype", line 5, in <module>
    from coldtype.renderer import main
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/coldtype/__init__.py", line 29, in <module>
    from coldtype.text import *
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/coldtype/text/__init__.py", line 2, in <module>
    from coldtype.text.reader import Style, StyledString, SegmentedString, normalize_font_path, Font, FontNotFoundException
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/coldtype/text/reader.py", line 26, in <module>
    from coldtype.fontgoggles.font.otfFont import OTFFont
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/coldtype/fontgoggles/font/otfFont.py", line 7, in <module>
    from ..misc.ftFont import FTFont
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/coldtype/fontgoggles/misc/ftFont.py", line 5, in <module>
    import freetype
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/freetype/__init__.py", line 24, in <module>
    from freetype.raw import *
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/freetype/raw.py", line 39, in <module>
    filename = ctypes.util.find_library('freetype')
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/coldtype/__init__.py", line 16, in find_library
    if res.startswith("/usr/local") and platform.processor() == "arm":
AttributeError: 'NoneType' object has no attribute 'startswith'
(animation-ct) mgpjm@MacBook-Pro-di-Mirko animation % coldtype demo
Traceback (most recent call last):
  File "/Users/mgpjm/animation/venv/bin/coldtype", line 5, in <module>
    from coldtype.renderer import main
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/coldtype/__init__.py", line 29, in <module>
    from coldtype.text import *
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/coldtype/text/__init__.py", line 2, in <module>
    from coldtype.text.reader import Style, StyledString, SegmentedString, normalize_font_path, Font, FontNotFoundException
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/coldtype/text/reader.py", line 26, in <module>
    from coldtype.fontgoggles.font.otfFont import OTFFont
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/coldtype/fontgoggles/font/otfFont.py", line 7, in <module>
    from ..misc.ftFont import FTFont
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/coldtype/fontgoggles/misc/ftFont.py", line 5, in <module>
    import freetype
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/freetype/__init__.py", line 24, in <module>
    from freetype.raw import *
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/freetype/raw.py", line 39, in <module>
    filename = ctypes.util.find_library('freetype')
  File "/Users/mgpjm/animation/venv/lib/python3.10/site-packages/coldtype/__init__.py", line 16, in find_library
    if res.startswith("/usr/local") and platform.processor() == "arm":
AttributeError: 'NoneType' object has no attribute 'startswith'
```

I tried to install al older version of skia-python as you suggest, but another error occur:
```
ERROR: Could not find a version that satisfies the requirement skia-python==86.1 (from versions: 87.4)
ERROR: No matching distribution found for skia-python==86.1
```
So I'm gettin little bit lost here, any suggestions?

Thank you!
m

**Reply by stenson** [reply]
Ah, sorry, yeah I should specify in the docs that python3.10 does not currently work with coldtype, since skia-python and freetype-py both don't yet support python3.10, both in arm and intel versions. If you are on an m1 computer, I'd recommend installing the latest python3.9 from python.org, and then when you build a virtual environment, build it with the `python3.9-intel64` command in the terminal (i.e. `python3.9-intel64 -m venv venv` then `source venv/bin/activate` and then you can use `python` directly and it'll be a 3.9 intel python that's fully compatible with all the required packages)

**Reply by mgpjm** [reply]
Thank you Rob, I switched to python3.9 and everything works just fine! 
I didn't have to specify `python3.9-intel64` anyway, just with `python3.9` seems be ok even if I'm on a m1 computer.

Anyway really super helpful like always
